python/src/iceberg/conversions.py

```diff
@@ -0,0 +1,267 @@
+            if partition_value_as_str is None:
+                return False
+            return partition_value_as_str.lower() == "true"
+
+Example (stacking registrations):
+   >>> @from_partition_value_to_py.register(FloatType)
+   >>> @from_partition_value_to_py.register(DoubleType)
+   >>> def _(primitive_type, partition_value_as_str: str):
+           return float(partition_value_as_str)
+"""
+import struct
+import uuid
+from decimal import ROUND_HALF_DOWN, Decimal, localcontext
+from functools import singledispatch
+
+from iceberg.types import (
+    BinaryType,
+    BooleanType,
+    DateType,
+    DecimalType,
+    DoubleType,
+    FixedType,
+    FloatType,
+    IntegerType,
+    LongType,
+    StringType,
+    TimestampType,
+    TimestamptzType,
+    TimeType,
+    UUIDType,
+)
+
+
+@singledispatch
+def from_partition_value_to_py(primitive_type, partition_value_as_str: str):
+    """A generic function function which converts a partition string to a python built-in
+
+    Args:
+        primitive_type(PrimitiveType): An implementation of the PrimitiveType base class
+        partition_value_as_str(str): A string representation of a partition value
+    """
+    if partition_value_as_str is None:
+        return None
+    raise TypeError(
+        f"Cannot convert partition string to python built-in, type {primitive_type} not supported: '{partition_value_as_str}'"
+    )
+
+
+@from_partition_value_to_py.register(BooleanType)
+def _(primitive_type, partition_value_as_str: str):
+    if partition_value_as_str is None:
+        return False
+    return partition_value_as_str.lower() == "true"
+
+
+@from_partition_value_to_py.register(IntegerType)
+@from_partition_value_to_py.register(LongType)
+def _(primitive_type, partition_value_as_str: str):
+    return int(float(partition_value_as_str))
+
+
+@from_partition_value_to_py.register(FloatType)
+@from_partition_value_to_py.register(DoubleType)
+def _(primitive_type, partition_value_as_str: str):
+    return float(partition_value_as_str)
+
+
+@from_partition_value_to_py.register(StringType)
+def _(primitive_type, partition_value_as_str: str):
+    return partition_value_as_str
+
+
+@from_partition_value_to_py.register(UUIDType)
+def _(primitive_type, partition_value_as_str: str):
+    return uuid.UUID(partition_value_as_str)
+
+
+@from_partition_value_to_py.register(FixedType)
+def _(primitive_type, partition_value_as_str: str):
+    return bytearray(bytes(partition_value_as_str, "UTF-8"))
+
+
+@from_partition_value_to_py.register(BinaryType)
+def _(primitive_type, partition_value_as_str: str):
+    return bytes(partition_value_as_str, "UTF-8")
+
+
+@from_partition_value_to_py.register(DecimalType)
+def _(primitive_type, partition_value_as_str: str):
+    return Decimal(partition_value_as_str)
+
+
+@singledispatch
+def to_bytes(primitive_type, value) -> bytes:
+    """A generic function which converts a built-in python value to bytes
```

Comment:
I think it would be good to refer to the spec here, since these implement single-value serialization: https://iceberg.apache.org/spec/#appendix-d-single-value-serialization

Reply:
Great idea, added this to the docstring:
```
This conversion follows the serialization scheme for storing single values as individual binary values defined in the Iceberg specification that
can be found at https://iceberg.apache.org/spec/#appendix-d-single-value-serialization
```